[Bug]: MTP Speculative Decoding with NVFP4: Weight Shape Mismatch

### Your current environment

==============================
        System Info
==============================
OS                           : Ubuntu 24.04.3 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version                : 18.1.3 (1ubuntu1)
CMake version                : version 3.28.3
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.10.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.3 (main, Jan 22 2026, 20:57:42) [GCC 13.3.0] (64-bit runtime)
Python platform              : Linux-6.14.0-37-generic-x86_64-with-glibc2.39

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.1.115
CUDA_MODULE_LOADING set to   :
GPU models and configuration :
GPU 0: NVIDIA RTX PRO 6000 Blackwell Workstation Edition
GPU 1: NVIDIA RTX 6000 Ada Generation

Nvidia driver version        : 580.126.09
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  24
On-line CPU(s) list:                     0-23
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) Ultra 9 285K
CPU family:                              6
Model:                                   198
Thread(s) per core:                      1
Core(s) per socket:                      24
Socket(s):                               1
Stepping:                                2
CPU(s) scaling MHz:                      118%
CPU max MHz:                             4600.0000
CPU min MHz:                             800.0000
BogoMIPS:                                7372.80
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni lam wbnoinvd dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid bus_lock_detect movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               768 KiB (20 instances)
L1i cache:                               1.3 MiB (20 instances)
L2 cache:                                40 MiB (12 instances)
L3 cache:                                36 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-23
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS Not affected; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.4
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.18.0
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.4.0
[pip3] nvidia-cutlass-dsl-libs-base==4.4.0
[pip3] nvidia-ml-py==13.590.48
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.4.5
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0
[pip3] torchaudio==2.10.0
[pip3] torchvision==0.25.0
[pip3] transformers==4.57.6
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.1.dev14100+g98b0205c3 (git sha: 98b0205c3)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
        GPU0    GPU1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PHB     0-23    0               N/A
GPU1    PHB      X      0-23    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_arivera

### 🐛 Describe the bug

**Model:** `nvidia/Qwen3-Next-80B-A3B-Instruct-NVFP4`
**vLLM versions tested:** v0.15.1, v0.16.0rc2.dev360+g2aab2bb54 (nightly)
**Hardware:** Single GPU

### Bug

When launching with `--speculative-config '{"method": "qwen3_next_mtp", "num_speculative_tokens": 2}'`, the engine crashes during drafter weight loading:
```
AssertionError: Shape mismatch: param=torch.Size([2048, 2048]) loaded=torch.Size([2048, 4096]) dtype=torch.bfloat16
```

Full traceback ends at:
```
File "vllm/model_executor/models/qwen3_next_mtp.py", line 217, in load_weights
    weight_loader(param, loaded_weight)
  File "vllm/model_executor/layers/linear.py", line 593, in weight_loader_v2
    param.load_column_parallel_weight(loaded_weight=loaded_weight)
  File "vllm/model_executor/parameter.py", line 153, in load_column_parallel_weight
    assert self.data.shape == loaded_weight.shape
```

---

### Root cause

`packed_modules_mapping` in `qwen3_next_mtp.py` line 230:
```python
# Current (wrong)
"gate_up_proj": ["up_proj", "down_proj"],

# Should be
"gate_up_proj": ["gate_proj", "up_proj"],
```

`down_proj` is not part of the fused gate+up projection. The wrong mapping causes the parameter to be allocated at half the required width (`[2048, 2048]`) while the checkpoint provides the correctly-sized fused weight (`[2048, 4096]`).

Note that lines 142-143 in the same file already have the correct shard mapping:
```python
("gate_up_proj", "gate_proj", 0),
("gate_up_proj", "up_proj", 1),
```
---

### Fix
```python
# qwen3_next_mtp.py line 230
"gate_up_proj": ["gate_proj", "up_proj"],
```

---

### To reproduce
```bash
docker run --rm --runtime nvidia --gpus 'device=0' \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  vllm/vllm-openai:v0.15.1-cu130 \
  --model nvidia/Qwen3-Next-80B-A3B-Instruct-NVFP4 \
  --attention-backend flashinfer \
  --kv-cache-dtype fp8 \
  --speculative-config '{"method": "qwen3_next_mtp", "num_speculative_tokens": 1}'
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

Confirming this is **still reproducible on `vllm-project/vllm` `main`** for **`nvidia/Qwen3-Next-80B-A3B-Instruct-NVFP4`** on a **DGX Spark (GB10, sm_121, 128 GB unified)** — building on @alkari's earlier analysis. Notably, `qwen3_next_mtp.py` is still **untouched**: the closed #35041 patched 6 other MTP files (deepseek/glm/step3/openpangu) but not `qwen3_next`, and #35675 / #39475 fix the sibling `qwen3_5_mtp.py`. So `qwen3_next` is currently the gap.

Three independent fixes were needed to get MTP loading + serving on this checkpoint:

**1. `packed_modules_mapping` copy-paste bug** (`qwen3_next_mtp.py`, still on `main`):
```python
# current (wrong):
"gate_up_proj": ["up_proj", "down_proj"],
# should match the base qwen3_next.py:
"gate_up_proj": ["gate_proj", "up_proj"],
```

**2. MTP `fc` head gets NVFP4-quantized → `AssertionError: param=[2048,2048] loaded=[2048,4096]`.** The ModelOpt checkpoint's quant-ignore glob `"mtp.layers.0*"` doesn't match `mtp.fc` / `mtp.norm` / `mtp.pre_fc_norm_*`, so `self.fc` (a quant-aware `ColumnParallelLinear`) tries to load the bf16 head as packed FP4. The clean, checkpoint-independent fix is the same one #35675 applied to `qwen3_5_mtp.py` — make the MTP `fc` unquantized:
```python
self.fc = ReplicatedLinear(self.config.hidden_size * 2, self.config.hidden_size,
                           bias=False, quant_config=None, prefix=f"{prefix}.fc")
# ... and in forward: hidden_states, _ = self.fc(hidden_states)
```
(A checkpoint-side workaround also works: add `"mtp.*"` to `quantization_config.ignore`, but the in-code fix avoids editing weights.)

**3. DGX Spark only — `pynvml` `NVML_ERROR_UNINITIALIZED`.** On unified memory, MTP init triggers a 2nd `nvmlShutdown()` (via `_resolve_auto_dtype` → `has_device_capability`) → `vllm/third_party/pynvml.py:_nvmlCheckReturn()` raises and aborts engine init. Tolerating `NVML_ERROR_UNINITIALIZED` there unblocks it. (Platform-specific; likely belongs at the call site rather than the vendored file.)

With all three applied (#1+#2 in a patched 0.23.0 image, #2 via the checkpoint workaround in our case, #3 in the image), the model **serves and generates correctly with the MTP drafter**, and MTP is a real win on GB10:

| config | c1 tok/s | c16 tok/s |
|---|---|---|
| baseline (no MTP) | 38.99 | 237.15 |
| MTP `num_speculative_tokens=1` | **50.63 (+29.9%)** | **266.52 (+12.4%)** |
| MTP `num_speculative_tokens=2` | 47.20 | 269.45 |

(`num_speculative_tokens=1` is best here — with a single nextn layer, `n>1` reuses it and lowers acceptance, matching vLLM's own warning. GuideLLM `concurrent`, chat 512/256, N=3 medians.)

Suggested path: extend the #35675 `fc → ReplicatedLinear(quant_config=None)` pattern to `qwen3_next_mtp.py` and apply the `gate_up_proj` mapping fix in the same PR; the `pynvml`/unified-memory item is a separate platform issue. Happy to share full repro/logs.

--- a/vllm/model_executor/models/qwen3_next_mtp.py
+++ b/vllm/model_executor/models/qwen3_next_mtp.py
@@ -64,13 +64,22 @@ class Qwen3NextMultiTokenPredictor(nn.Module):
             config.hidden_size,
         )
 
+        # Workaround: mtp.fc is stored as BF16 in NVFP4 checkpoints but is
+        # missing from the checkpoint quant exclude list (its `ignore` glob
+        # does not cover `mtp.fc`). Force unquantized to match the weights,
+        # mirroring the Qwen3.5 MTP handling (PR #38832).
+        fc_quant = (
+            None
+            if (quant_config and quant_config.get_name() == "modelopt_fp4")
+            else quant_config
+        )
         self.fc = ColumnParallelLinear(
             self.config.hidden_size * 2,
             self.config.hidden_size,
             gather_output=True,
             bias=False,
             return_bias=False,
-            quant_config=quant_config,
+            quant_config=fc_quant,
             prefix=f"{prefix}.fc",
         )
 
@@ -242,7 +251,7 @@ class Qwen3NextMTP(nn.Module, QwenNextMixtureOfExperts):
             "k_proj",
             "v_proj",
         ],
-        "gate_up_proj": ["up_proj", "down_proj"],
+        "gate_up_proj": ["gate_proj", "up_proj"],
     }
 
     def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):

